### PR TITLE
[Tests] Test attention.py

### PR DESCRIPTION
**Pull request by sidthekidder**
Added blackbox tests for AttentionBlock and SpatialTransformer - #201. A question - since BasicTransformerBlock, CrossAttention and FeedForward classes are all used within SpatialTransformerBlock, is it required to test them separately? Or testing SpatialTransformer is enough, since any change in them would cause the SpatialTransformerBlockTests to fail.

### CODE REVIEW

File: src/diffusers/models/__init__.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from .attention import AttentionBlock, SpatialTransformer
```
Comment (patil-suraj) [inline src/diffusers/models/__init__.py:15]: This shouldn't be exposed here.
```suggestion
```

File: tests/test_layers_utils.py [hunk 2]
```diff
@@ -216,3 +217,45 @@ def test_downsample_with_conv_out_dim(self):
...
+class SpatialTransformerTests(unittest.TestCase):
+    def test_spatial_transformer_default(self):
+        torch.manual_seed(0)
+        sample = torch.randn(1, 32, 64, 64)
+        spatialTransformerBlock = SpatialTransformer(
+            in_channels=32,
+            n_heads=1,
+            d_head=32,
+            dropout=0.0,
+            context_dim=None,
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:252]: We should test all possible cases here. Both with `context_dim=None` and `context_dim=some_value`

File: tests/test_layers_utils.py [hunk 3]
```diff
@@ -216,3 +217,45 @@ def test_downsample_with_conv_out_dim(self):
...
+        )
+        with torch.no_grad():
+            attention_scores = spatialTransformerBlock(sample)
+
+        assert attention_scores.shape == (1, 32, 64, 64)
+        output_slice = attention_scores[0, -1, -3:, -3:]
+
+        expected_slice = torch.tensor([-1.2447, -0.0137, -0.9559, -1.5223, 0.6991, -1.0126, -2.0974, 0.8921, -1.0201])
+        assert torch.allclose(output_slice.flatten(), expected_slice, atol=1e-1)
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:261]: tol of `1e-1` is too low, think it should pass with 1e-3 here.

File: tests/test_layers_utils.py [hunk 4]
```diff
@@ -216,3 +217,45 @@ def test_downsample_with_conv_out_dim(self):
...
+    def test_attention_block_default(self):
+        torch.manual_seed(0)
+        sample = torch.randn(1, 32, 64, 64)
+        attentionBlock = AttentionBlock(
+            channels=32,
+            num_head_channels=1,
+            rescale_output_factor=1.0,
+            eps=1e-6,
+            num_groups=32,
+        )
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:232]: we should handle device here, for example as it's done in other tests

https://github.com/huggingface/diffusers/blob/f085d2f5c6569a1c0d90327c51328622036ef76e/tests/test_models_unet.py#L139-L150

File: tests/test_layers_utils.py [hunk 5]
```diff
@@ -216,3 +217,45 @@ def test_downsample_with_conv_out_dim(self):
...
+
+class SpatialTransformerTests(unittest.TestCase):
+    def test_spatial_transformer_default(self):
+        torch.manual_seed(0)
+        sample = torch.randn(1, 32, 64, 64)
+        spatialTransformerBlock = SpatialTransformer(
+            in_channels=32,
+            n_heads=1,
+            d_head=32,
+            dropout=0.0,
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:251]: same comment about the device.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for working this ! Left some comments. Let me know if you need any help with it :)

File: tests/test_layers_utils.py (hunk 2 above)
Comment (sidthekidder) [reply tests/test_layers_utils.py:259]: Added test for multiple values of `context_dim` and `dropout` as well.

**Comment by sidthekidder** [comment]
Thanks for the helpful comments @patil-suraj! Updated to address them

**Comment by sidthekidder** [comment]
Hmm it seems the dropout test is failing maybe because nn.Dropout is probabilistic, though it was passing locally for me consistently.

**Comment by sidthekidder** [comment]
@patil-suraj I removed a test that was failing (nn.dropout which had different results on my local and the CI server). Any other suggestions

**Comment by sidthekidder** [comment]
@patil-suraj does this PR look ok?

File: tests/test_layers_utils.py [hunk 6]
```diff
@@ -216,3 +218,74 @@ def test_downsample_with_conv_out_dim(self):
...
+
+
+class SpatialTransformerTests(unittest.TestCase):
+    def test_spatial_transformer_default(self):
+        torch.manual_seed(0)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(0)
+
+        sample = torch.randn(1, 32, 64, 64).to(torch_device)
+        spatialTransformerBlock = SpatialTransformer(
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:254]:
```suggestion
        spatial_transformer_block = SpatialTransformer(
```

File: tests/test_layers_utils.py [hunk 7]
```diff
@@ -216,3 +218,74 @@ def test_downsample_with_conv_out_dim(self):
...
+            in_channels=32,
+            n_heads=1,
+            d_head=32,
+            dropout=0.0,
+            context_dim=None,
+        ).to(torch_device)
+        with torch.no_grad():
+            attention_scores = spatialTransformerBlock(sample)
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:262]:
```suggestion
            attention_scores = spatial_transformer_block(sample)
```

File: tests/test_layers_utils.py [hunk 8]
```diff
@@ -216,3 +218,74 @@ def test_downsample_with_conv_out_dim(self):
...
+        assert torch.allclose(output_slice.flatten(), expected_slice, atol=1e-3)
+
+    def test_spatial_transformer_context_dim(self):
+        torch.manual_seed(0)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(0)
+
+        torch.manual_seed(0)
+        sample = torch.randn(1, 64, 64, 64).to(torch_device)
+        spatialTransformerBlock = SpatialTransformer(
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:277]:
```suggestion
        spatial_transformer_block = SpatialTransformer(
```

File: tests/test_layers_utils.py [hunk 9]
```diff
@@ -216,3 +218,74 @@ def test_downsample_with_conv_out_dim(self):
...
+            in_channels=64,
+            n_heads=2,
+            d_head=32,
+            dropout=0.0,
+            context_dim=64,
+        ).to(torch_device)
+        with torch.no_grad():
+            attention_scores = spatialTransformerBlock(sample)
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:285]:
```suggestion
            attention_scores = spatial_transformer_block(sample)
```

Comment (patil-suraj) [inline tests/test_layers_utils.py:285]: Here we should also pass the `context` argument along with the `sample`, it could be created as follows

```python
context = torch.randn(1, 4, 64).to(torch_device)
```

File: tests/test_layers_utils.py [hunk 10]
```diff
@@ -216,3 +218,74 @@ def test_downsample_with_conv_out_dim(self):
...
+
+        expected_slice = torch.tensor([-1.2447, -0.0137, -0.9559, -1.5223, 0.6991, -1.0126, -2.0974, 0.8921, -1.0201])
+        assert torch.allclose(output_slice.flatten(), expected_slice, atol=1e-3)
+
+    def test_spatial_transformer_context_dim(self):
+        torch.manual_seed(0)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(0)
+
+        torch.manual_seed(0)
```
Comment (patil-suraj) [inline tests/test_layers_utils.py:275]: It's already called above.
```suggestion
```

**Review by patil-suraj — COMMENTED** [review]
Just left couple of nits, then it should be good for merge.

**Comment by patil-suraj** [comment]
> @patil-suraj I removed a test that was failing (nn.dropout which had different results on my local and the CI server). Any other suggestions

Aah, I see. We should actually put the blocks in `eval` mode. This will disable the dropout.

File: tests/test_layers_utils.py (hunk 9 above)
Comment (sidthekidder) [reply tests/test_layers_utils.py]: Oh thanks, missed that parameter!

Comment (sidthekidder) [reply tests/test_layers_utils.py]: Updated name everywhere

**Comment by sidthekidder** [comment]
Thanks for the eval tip, that worked! Added a test with `dropout=0.3` as well.

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this!

File: tests/test_layers_utils.py [hunk 11]
```diff
@@ -216,3 +218,100 @@ def test_downsample_with_conv_out_dim(self):
...
+
+        assert attention_scores.shape == (1, 32, 64, 64)
+        output_slice = attention_scores[0, -1, -3:, -3:]
+
+        expected_slice = torch.tensor([-1.2447, -0.0137, -0.9559, -1.5223, 0.6991, -1.0126, -2.0974, 0.8921, -1.0201])
+        assert torch.allclose(output_slice.flatten(), expected_slice, atol=1e-3)
+
+    def test_spatial_transformer_context_dim(self):
+        torch.manual_seed(0)
+        if torch.cuda.is_available():
```
Comment (patrickvonplaten) [inline tests/test_layers_utils.py:272]: (nit) don't think this is needed anymore actually in newer pytorch versions

**Review by patrickvonplaten — COMMENTED** [review]
Very nice!
